### PR TITLE
Unsatisfiable side condition involving `\forall`: add `translatePredicateForall`

### LINKED ISSUE #2138: Unsatisfiable side condition involving `\forall` not correctly checked
When executing a program with `\forall` as side condition, a state with an unsatisfiable side condition is not pruned.
I think there is something wrong with the smt encoding, but I'm not sure.

Definition:

```k
module TEST
    imports BOOL
    imports INT

    syntax Exp ::= Int | Bool
                 | pair(Int, Int) [klabel(pair), symbol]

    syntax KItem ::= "#assume" Exp [klabel(assume), symbol]
    rule #assume true => .K
    rule #assume false => #Bottom

    syntax KItem ::= "#assert" Exp  [klabel(assert), symbol]
                   | "#fail"
    rule #assert true => .K
    rule #assert false => #fail

// Uninterpreted functions
    syntax Int ::= i(Int, Int)   [function, functional, no-evaluators, smtlib(fi), klabel(i)]
    syntax Int ::= p(Exp, Int)   [function, functional, no-evaluators, smtlib(fp), klabel(p)]
endmodule
```

This program:

```
  ( #Forall figVar?I:Int . {
    i ( figVar?I:Int , 0 ) ==Int 0
  #Equals
    true
  } )
#And
  <k>
    #assert i ( 1 , 0 ) ==Int 0 ~> .
  </k>
```

results in:

```
    {
      false
    #Equals
      i ( 1 , 0 ) ==Int 0
    }
  #And
    ( #Forall figVar?I:Int . {
      i ( figVar?I:Int , 0 )
    #Equals
      0
    } )
  #And
    {
      Result:GeneratedTopCell
    #Equals
      <k>
        #fail ~> .
      </k>
    }
#Or
    {
      true
    #Equals
      i ( 1 , 0 ) ==Int 0
    }
  #And
    ( #Forall figVar?I:Int . {
      i ( figVar?I:Int , 0 )
    #Equals
      0
    } )
  #And
    {
      Result:GeneratedTopCell
    #Equals
      <k>
        .
      </k>
    }
```

Note that the first branch is unsatisfiable.

The (handwritten) kore program is below (and bug report attached):

```
\and {SortGeneratedTopCell{}}
     ( \forall {SortGeneratedTopCell{}} ( ConfigVar'Ques'I:SortInt{}
                                        , \equals{SortBool{}, SortGeneratedTopCell{}} (
                                               Lbl'UndsEqlsEqls'Int'Unds'{}( Lbli'LParUndsCommUndsRParUnds'TEST'Unds'Int'Unds'Int'Unds'Int{}( ConfigVar'Ques'I:SortInt{}, \dv{SortInt{}}("0") ) , \dv{SortInt{}}("0") )
                                             , \dv{SortBool{}}("true")
                                             )
                                        )
     ,  Lbl'-LT-'generatedTop'-GT-'{}(
            Lbl'-LT-'k'-GT-'{}(
                    kseq{}(
                        Lblassert{}(
                             inj{SortBool{}, SortExp{}}(
                                Lbl'UndsEqlsEqls'Int'Unds'{}(
                                    Lbli'LParUndsCommUndsRParUnds'TEST'Unds'Int'Unds'Int'Unds'Int{}(
                                         \dv{SortInt{}}("1"),
                                         \dv{SortInt{}}("0")
                                    ),
                                     \dv{SortInt{}}("0")
                                )
                            )
                        ),
                         dotk{}()
                    )
            ),
            Lbl'-LT-'generatedCounter'-GT-'{}( \dv{SortInt{}}("0") )
        )
    )    
```
[kore-exec.tar.gz](https://github.com/kframework/kore/files/5215815/kore-exec.tar.gz)

### PR DESCRIPTION
**Pull request by andreiburdusa**
---
Fixes #2138 
###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`

### CODE REVIEW

**Comment by andreiburdusa** [comment]
For the example given in the bug report from the issue, the result is no more
```
/* Spa */
\or{SortGeneratedTopCell{}}(
    /* Created: Kore.Internal.Conditional.makeAndPredicate */
    /* Spa */
    \and{SortGeneratedTopCell{}}(
        /* Created: Kore.Internal.Predicate.makeAndPredicate */
        /* Sfa */
        \and{SortGeneratedTopCell{}}(
            /* Sfa */
            \equals{SortBool{}, SortGeneratedTopCell{}}(
                /* Fl Fn D Sfa Cl */ /* builtin: */ \dv{SortBool{}}("false"),
                /* Fl Fn D Sfa */
                Lbl'UndsEqlsEqls'Int'Unds'{}(
                    /* Fl Fn D Sfa */
                    Lbli'LParUndsCommUndsRParUnds'TEST'Unds'Int'Unds'Int'Unds'Int{}(
                        /* Fl Fn D Sfa Cl */ /* builtin: */ \dv{SortInt{}}("1"),
                        /* Fl Fn D Sfa Cl */ /* builtin: */ \dv{SortInt{}}("0")
                    ),
                    /* Fl Fn D Sfa Cl */ /* builtin: */ \dv{SortInt{}}("0")
                )
            ),
            /* Sfa */
            \forall{SortGeneratedTopCell{}}(
                ConfigVar'Ques'I:SortInt{},
                /* Sfa */
                \equals{SortInt{}, SortGeneratedTopCell{}}(
                    /* Fl Fn D Sfa */
                    Lbli'LParUndsCommUndsRParUnds'TEST'Unds'Int'Unds'Int'Unds'Int{}(
                        /* Fl Fn D Sfa */ ConfigVar'Ques'I:SortInt{},
                        /* Fl Fn D Sfa Cl */ /* builtin: */ \dv{SortInt{}}("0")
                    ),
                    /* Fl Fn D Sfa Cl */ /* builtin: */ \dv{SortInt{}}("0")
                )
            )
        ),
        /* Spa */
        \equals{SortGeneratedTopCell{}, SortGeneratedTopCell{}}(
            /* Fl Fn D Sfa */ VarResult:SortGeneratedTopCell{},
            /* Fl Fn D Sfa Cl */
            Lbl'-LT-'generatedTop'-GT-'{}(
                /* Fl Fn D Sfa Cl */
                Lbl'-LT-'k'-GT-'{}(
                    /* Fl Fn D Sfa Cl */
                    kseq{}(
                        /* Fl Fn D Sfa Cl */
                        Lbl'Hash'fail'Unds'TEST'Unds'KItem{}(),
                        /* Fl Fn D Sfa Cl */ dotk{}()
                    )
                ),
                /* Fl Fn D Sfa Cl */
                Lbl'-LT-'generatedCounter'-GT-'{}(
                    /* Fl Fn D Sfa Cl */ /* builtin: */ \dv{SortInt{}}("0")
                )
            )
        )
    ),
    /* Created: Kore.Internal.Conditional.makeAndPredicate */
    /* Spa */
    \and{SortGeneratedTopCell{}}(
        /* Created: Kore.Internal.Predicate.makeAndPredicate */
        /* Sfa */
        \and{SortGeneratedTopCell{}}(
            /* Sfa */
            \equals{SortBool{}, SortGeneratedTopCell{}}(
                /* Fl Fn D Sfa Cl */ /* builtin: */ \dv{SortBool{}}("true"),
                /* Fl Fn D Sfa */
                Lbl'UndsEqlsEqls'Int'Unds'{}(
                    /* Fl Fn D Sfa */
                    Lbli'LParUndsCommUndsRParUnds'TEST'Unds'Int'Unds'Int'Unds'Int{}(
                        /* Fl Fn D Sfa Cl */ /* builtin: */ \dv{SortInt{}}("1"),
                        /* Fl Fn D Sfa Cl */ /* builtin: */ \dv{SortInt{}}("0")
                    ),
                    /* Fl Fn D Sfa Cl */ /* builtin: */ \dv{SortInt{}}("0")
                )
            ),
            /* Sfa */
            \forall{SortGeneratedTopCell{}}(
                ConfigVar'Ques'I:SortInt{},
                /* Sfa */
                \equals{SortInt{}, SortGeneratedTopCell{}}(
                    /* Fl Fn D Sfa */
                    Lbli'LParUndsCommUndsRParUnds'TEST'Unds'Int'Unds'Int'Unds'Int{}(
                        /* Fl Fn D Sfa */ ConfigVar'Ques'I:SortInt{},
                        /* Fl Fn D Sfa Cl */ /* builtin: */ \dv{SortInt{}}("0")
                    ),
                    /* Fl Fn D Sfa Cl */ /* builtin: */ \dv{SortInt{}}("0")
                )
            )
        ),
        /* Spa */
        \equals{SortGeneratedTopCell{}, SortGeneratedTopCell{}}(
            /* Fl Fn D Sfa */ VarResult:SortGeneratedTopCell{},
            /* Fl Fn D Sfa Cl */
            Lbl'-LT-'generatedTop'-GT-'{}(
                /* Fl Fn D Sfa Cl */
                Lbl'-LT-'k'-GT-'{}(/* Fl Fn D Sfa Cl */ dotk{}()),
                /* Fl Fn D Sfa Cl */
                Lbl'-LT-'generatedCounter'-GT-'{}(
                    /* Fl Fn D Sfa Cl */ /* builtin: */ \dv{SortInt{}}("0")
                )
            )
        )
    )
)
```
but 
```
/* Created: Kore.Internal.Conditional.makeAndPredicate */
/* Spa */
\and{SortGeneratedTopCell{}}(
    /* Created: Kore.Internal.Predicate.makeAndPredicate */
    /* Sfa */
    \and{SortGeneratedTopCell{}}(
        /* Sfa */
        \equals{SortBool{}, SortGeneratedTopCell{}}(
            /* Fl Fn D Sfa Cl */ /* builtin: */ \dv{SortBool{}}("true"),
            /* Fl Fn D Sfa */
            Lbl'UndsEqlsEqls'Int'Unds'{}(
                /* Fl Fn D Sfa */
                Lbli'LParUndsCommUndsRParUnds'TEST'Unds'Int'Unds'Int'Unds'Int{}(
                    /* Fl Fn D Sfa Cl */ /* builtin: */ \dv{SortInt{}}("1"),
                    /* Fl Fn D Sfa Cl */ /* builtin: */ \dv{SortInt{}}("0")
                ),
                /* Fl Fn D Sfa Cl */ /* builtin: */ \dv{SortInt{}}("0")
            )
        ),
        /* Sfa */
        \forall{SortGeneratedTopCell{}}(
            ConfigVar'Ques'I:SortInt{},
            /* Sfa */
            \equals{SortInt{}, SortGeneratedTopCell{}}(
                /* Fl Fn D Sfa */
                Lbli'LParUndsCommUndsRParUnds'TEST'Unds'Int'Unds'Int'Unds'Int{}(
                    /* Fl Fn D Sfa */ ConfigVar'Ques'I:SortInt{},
                    /* Fl Fn D Sfa Cl */ /* builtin: */ \dv{SortInt{}}("0")
                ),
                /* Fl Fn D Sfa Cl */ /* builtin: */ \dv{SortInt{}}("0")
            )
        )
    ),
    /* Spa */
    \equals{SortGeneratedTopCell{}, SortGeneratedTopCell{}}(
        /* Fl Fn D Sfa */ VarResult:SortGeneratedTopCell{},
        /* Fl Fn D Sfa Cl */
        Lbl'-LT-'generatedTop'-GT-'{}(
            /* Fl Fn D Sfa Cl */
            Lbl'-LT-'k'-GT-'{}(/* Fl Fn D Sfa Cl */ dotk{}()),
            /* Fl Fn D Sfa Cl */
            Lbl'-LT-'generatedCounter'-GT-'{}(
                /* Fl Fn D Sfa Cl */ /* builtin: */ \dv{SortInt{}}("0")
            )
        )
    )
)
```
It doesnt have the unsatisfiable branch anymore

**Review by ttuegel — CHANGES_REQUESTED** [review]
For the integration test, please check in the K definition and spec instead. Please do not check in the SMT transcript.

File: test/issue-2138/issue-2138-spec.k [hunk 1]
```diff
@@ -0,0 +1,12 @@
+module VERIFICATION
+  import TEST
+endmodule
+
+module ISSUE-2138-SPEC
+  import VERIFICATION
+
+  rule
+    <k> #assert i ( 1 , 0 ) ==Int 0 => . </k>
+      requires #assert i ( X, 0 ) ==Int 0
```
Comment (ttuegel) [inline test/issue-2138/issue-2138-spec.k:10]: This claim is non-sensical. `#assert` isn't a function, but a constructor.

**Review by ttuegel — CHANGES_REQUESTED** [review]
The integration test must have a `#Forall`.

File: test/issue-2138/issue-2138-spec.k (hunk 1 above)
[the remark this answers is not in the document]
Comment (ttuegel) [reply test/issue-2138/issue-2138-spec.k:10]: Please open a bug with the frontend.

**Review by ttuegel — APPROVED** [review]

**Pull request merged**